examples/tqc/utils.py

```diff
@@ -0,0 +1,410 @@
+    abs_pairwise_delta = torch.abs(pairwise_delta)
+    huber_loss = torch.where(abs_pairwise_delta > 1,
+                             abs_pairwise_delta - 0.5,
+                             pairwise_delta ** 2 * 0.5)
+    n_quantiles = quantiles.shape[2]
+    tau = torch.arange(n_quantiles, device=quantiles.device).float() / n_quantiles + 1 / 2 / n_quantiles
+    loss = (torch.abs(tau[None, None, :, None] - (pairwise_delta < 0).float()) * huber_loss).mean()
+    return loss
+
+
+# ====================================================================
+# TQC Loss
+# --------
+
+class TQCLoss(LossModule):
+    def __init__(
+            self,
+            actor_network,
+            qvalue_network,
+            gamma,
+            top_quantiles_to_drop,
+            alpha_init,
+            device
+    ):
+        super(type(self), self).__init__()
+        super().__init__()
+
+        self.convert_to_functional(
+            actor_network,
+            "actor",
+            create_target_params=False,
+            funs_to_decorate=["forward", "get_dist"],
+        )
+
+        self.convert_to_functional(
+            qvalue_network,
+            "critic",
+            create_target_params=True  # Create a target critic network
+        )
+
+        self.device = device
+        self.log_alpha = torch.tensor([np.log(alpha_init)], requires_grad=True, device=self.device)
+        self.gamma = gamma
+        self.top_quantiles_to_drop = top_quantiles_to_drop
+
+        # Compute target entropy
+        action_spec = getattr(self.actor, "spec", None)
+        if action_spec is None:
+            print("Could not deduce action spec from actor network.")
+        if not isinstance(action_spec, CompositeSpec):
+            action_spec = CompositeSpec({"action": action_spec})
+        action_container_len = len(action_spec.shape)
+        self.target_entropy = -float(action_spec["action"].shape[action_container_len:].numel())
+
+    def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
+        td_next = tensordict["next"]
+        reward = td_next["reward"]
+        not_done = tensordict["done"].logical_not()
+
+        alpha = torch.exp(self.log_alpha)
+
+        # Q-loss
+        with torch.no_grad():
+            # get policy action
+            td_next = self.actor(td_next, params=self.actor_params)
+            td_next = self.critic(td_next, params=self.target_critic_params)
+
+            next_log_pi = td_next["sample_log_prob"]
+            next_log_pi = torch.unsqueeze(next_log_pi, dim=-1)
+
+            # compute and cut quantiles at the next state
+            next_z = td_next["state_action_value"]
+            sorted_z, _ = torch.sort(next_z.reshape(next_z.shape[0], -1))
+            sorted_z_part = sorted_z[:, :-self.top_quantiles_to_drop]
+
+            # compute target
+            target = reward + not_done * self.gamma * (sorted_z_part - alpha * next_log_pi)
```

Comment:
Can we use our value estimators for this?

Reply:
There is a slight issue / inconvenience with this: the shape of reward is [batch size, 1], whereas the shape of the term in the bracket on the right side is [batch size, n_nets * n_quantiles]. In the computation of target, there is therefore some broadcasting going on, and target ends up having shape [batch size, n_nets * n_quantiles]. The critic outputs a tensor of shape [batch size, n_nets, n_quantiles] (although this could be reshaped to [batch size, n_nets * n_quantiles] of course). However, the TD0Estimator (or any other one presumably) expects "reward" and "state_action_value" to be of the same shape (i.e. it does not support broadcasting in the computation of target)

Reply:
Not sure what a reasonable solution here would be. The core issue seems to be that TD0Estimator does not allow for broadcasting in the computation of target; however we want to broadcast here since the critic doesn't output one value estimate but a collection of quantile estimates for the q value function.

Reply:
The error arising from this issue (critic output of different shape than reward) is:

`RuntimeError: All input tensors (value, reward and done states) must share a unique shape.`

Reply:
I have now added some code to demonstrate this problem in the latest commit (it's a bit too cumbersome to post all the code in a comment). Running it should reproduce the issue .

Reply:
Sorry for leaving this dormant for so long
It's ok like this if the value can't be adapted. I would just have a dummy `make_value_estimator` that does nothing else than checking the value estimator type and telling the user if that doesn't fit what is required. (no need to set it since the code takes care of it)

Reply:
No worries! Ok - I'll see to doing that on Monday 😊 

Do you think it would make sense to allow the value estimators to accept critic outputs of a different shape? (Like is needed for TQC) What's the reason for imposing the restriction on critic shape?

Reply:
Done. Would you like me to move the TQCLoss implementation into the objectives directory?

Reply:
Sounds good!

That makes sense, thanks for explaining! Perhaps one could add an option to explicitly specify the shapes of reward, value and done states when constructing the estimator? Then do some consistency checks on those shapes.